src/xinspect.hpp

```diff
@@ -199,12 +196,84 @@ namespace xeus
 
                     if (!node.empty())
                     {
-                       return url + node;
+                       inspect_result = url + node;
                     }
                 }
             }
         }
-        return "";
+
+        std::cout << "inspect_result " << inspect_result << "\n";
+        if (inspect_result.empty())
+        {
+            std::cerr << "No documentation found for " << code << "\n";
+            std::cout << std::flush;
+            kernel_res["found"] = false;
+            kernel_res["status"] = "error";
+            kernel_res["ename"] = "ename";
+            kernel_res["evalue"] = "evalue";
+            kernel_res["traceback"] = {};
+        }
+        else
+        {
+            // Format html content.
+            std::string html_content = R"(<style>
+            #pager-container {
+                padding: 0;
+                margin: 0;
+                width: 100%;
+                height: 100%;
+            }
+            .xeus-iframe-pager {
+                padding: 0;
+                margin: 0;
+                width: 100%;
+                height: 100%;
+                border: none;
+            }
+            </style>
+            <iframe class="xeus-iframe-pager" src=")" + inspect_result + R"("></iframe>)";
+
+            kernel_res["payload"] = {
+                xjson::object({
+                    {"data", {
+                        {"text/plain", inspect_result},
+                        {"text/html", html_content}}},
+                    {"source", "page"},
+                    {"start", 0}})
+            };
+
+            kernel_res["data"] =
+                xjson::object({{"text/plain", inspect_result},
+                               {"text/html", html_content}})
+            ;
+
+            std::cout << std::flush;
+            kernel_res["found"] = true;
+            kernel_res["status"] = "ok";
+        }
+
     }
+
+    struct xintrospection: xpreamble
+    {
+        const std::string spattern = R"(^\?)";
```

Comment:
Could the spatter be a static constant member, remove the xsystem and xmagic constructors completely and make use of it directly instead of copying it into a `pattern` member?

Reply:
I am not sure that it is possible because I need to call set_pattern.

Reply:
why? this all seems to be redundant.

Reply:
I don't see the need to a smart pointer. A value semantics would be much simpler. The pattern can be defined in the derived class as a const static variable.